mediatek: fix nmbm configuration mismatch (Xiaomi AX3000t)

csharper2005 opened the pull request:
- kernel: nmbm: add mediatek,bmt-mtd-overridden-oobsize property
- mediatek: fix nmbm configuration mismatch (Xiaomi AX3000t)

DragonBluep commented:
Medaitek u-boot defines oob size as 64 bytes. Therefore, I think it makes more sense to add a hack patch on the Mediatek target to overwrite oob to 64 bytes.

csharper2005 commented:
> Medaitek u-boot defines oob size as 64 bytes. Therefore, I think it makes more sense to add a hack patch on the Mediatek target to overwrite oob to 64 bytes.

Maybe. In the current opensource u-boot this chip has 64 bytes oob too: https://github.com/openwrt/openwrt/blob/e9415be4ad3cacbbd6e65fee5213baf3eaf9e07a/package/boot/uboot-mediatek/patches/100-22-mtd-spi-nand-backport-from-upstream-kernel.patch#L514

But we have other chips, with 128 bytes oob in mediatek u-boot target. Won't there be a negative impact? 
https://github.com/openwrt/openwrt/blob/e9415be4ad3cacbbd6e65fee5213baf3eaf9e07a/package/boot/uboot-mediatek/patches/100-22-mtd-spi-nand-backport-from-upstream-kernel.patch#L530

Can you explain more your idea? What should this hack patch be applied to?

robimarko commented:
This NAND is not in the upstream U-Boot, it's just added via MTK feed patch.

64B OOB is just not correct, look at the NAND datasheet

csharper2005 commented:
@robimarko, ok, 64B OOB isn't correct. Xiaomi u-boot and firmware use this incorrect value. OpenWrt users have this incorrect u-boot installed on their devices. And we need make nmbm working. What do you suggest?

csharper2005 commented:
Hack patch works.
```
Sun Jan 12 13:58:09 2025 kern.info kernel: [    0.871732] spi-nand spi0.0: Winbond SPI NAND was found.
Sun Jan 12 13:58:09 2025 kern.info kernel: [    0.877049] spi-nand spi0.0: 128 MiB, block size: 128 KiB, page size: 2048, OOB size: 64
Sun Jan 12 13:58:09 2025 kern.info kernel: [    0.885711] Signature found at block 1023 [0x07fe0000]
Sun Jan 12 13:58:09 2025 kern.info kernel: [    0.890866] NMBM management region starts at block 960 [0x07800000]
Sun Jan 12 13:58:09 2025 kern.info kernel: [    0.897936] First info table with writecount 1 found in block 960
Sun Jan 12 13:58:09 2025 kern.info kernel: [    0.906335] Second info table with writecount 1 found in block 963
```

csharper2005 commented:
https://github.com/openwrt/openwrt/commit/ab95c402dcc0072af69f743c7389f1036792f053:
- Nmbm changes were replaced by the hack patch.

hauke commented on target/linux/mediatek/patches-6.6/108-hack-adjust-Winbond-W25N01KV-spi-nand-oob.patch:
```diff
@@ -0,0 +1,11 @@
+--- a/drivers/mtd/nand/spi/winbond.c
++++ b/drivers/mtd/nand/spi/winbond.c
+@@ -182,7 +182,7 @@ static const struct spinand_info winbond
```
Please add a patch description. This should contain why this is needed.

csharper2005 replied:
Done.

csharper2005 commented:
https://github.com/openwrt/openwrt/commit/13ec9e694433fc21ddc7a884e20ed67765bbce1d:
- Add description to the patch file

sc16me commented:
Tested-by: Aleksandr Danilov sc16me@gmail.com

csharper2005 commented:
@robimarko, @hauke Please let me know, If we need as in the datasheet and don't care how it works on real devices. In this case I will close the PR and attach the patch to the issue. Affected users can build the firmware themselves.

csharper2005 commented:
> 64B OOB is just not correct, look at the NAND datasheet

Today I looked at the 25N01KVZEI**R** datasheet (_W25N01KVxxx**R**_Datasheet_Rev.H.pdf_) and found that **64B OOB is correct**:
![image](https://github.com/user-attachments/assets/8acff89a-d5e8-4657-95c9-8104f98d7fb4)

![image](https://github.com/user-attachments/assets/9bdbbb41-e70d-4b84-8298-9ed933488126)


Chip photo:
![image](https://github.com/user-attachments/assets/1de9e081-c812-4033-a627-a86ab536e340)

robimarko commented:
User area is 64B, but there is also parity area which is 32B, hence 96B.

This is the logic with previous Winbond IC-s so I followed the same, larger models follow th same logic as well

csharper2005 commented:
https://www.winbond.com/hq/product/code-storage-flash-memory/qspinand-flash/?__locale=en&partNo=W25N01KV
1Gb Serial NAND Flash Memory with uniform **2KB+64B** page size. (**2KB+96B** page size for Sequential Read Mode)

So what do we do with this? How do we fix nmbm? ☝️ 👀

DragonBluep commented:
@hackpascal Hi! Could you please help to review this PR?

hauke commented:
I perfere a DTS option. We can active this option only for this device. If some other device uses this flash chip we can would use the correct OOB size.

hackpascal commented:
> Will there be a threat to the integrity of the data

OOB is unused in UBI and only 0xff data will be write to spi-nand.
This will not affect the parity data of the main data area.

csharper2005 committed:
kernel: nmbm: add mediatek,bmt-mtd-overridden-oobsize property

This commit adds new "mediatek,bmt-mtd-overridden-oobsize" property. The
property helps avoid "NMBM configuration mismatch" error if mtd "OOB size"
is not equal to the "spare size" which is stored in the nmbm signature.

Signed-off-by: Mikhail Zhilkin <csharper2005@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17549
Signed-off-by: Robert Marko <robimarko@gmail.com>

csharper2005 committed:
mediatek: fix nmbm configuration mismatch (Xiaomi AX3000t)

This commit fixes nmbm configuration mismatch error on Xiaomi AX3000t
with Winbond W25N01KVZEIR spi-nand:
'''
[    0.786783] NMBM configuration mismatch
'''

Root cause:
1. U-Boot W25N01KV spi-nand driver ia compiled with 64B OOB size for the
   chip and store this size in the nmbm signature;
2. Linux W25N01KV driver use 96B OOB.

The change doesn't affect AX3000t variants with other spi-nand chips
(ESMT, Foresee) because their Linux drivers use 64B OOB.

Fixes: openwrt#16972
Tested-by: Aleksandr Danilov <sc16me@gmail.com>
Signed-off-by: Mikhail Zhilkin <csharper2005@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17549
Signed-off-by: Robert Marko <robimarko@gmail.com>

csharper2005 commented:
https://github.com/openwrt/openwrt/commit/5875aeafe6e2e44f61eeeae4f6b0d19371497f29:
Update PR according to the latest suggestions:
- kernel: nmbm: add mediatek,bmt-mtd-overridden-oobsize property
- mediatek: fix nmbm configuration mismatch (Xiaomi AX3000t)

csharper2005 commented:
24.10.0 with patch (if there are people willing to test it):
https://mega.nz/folder/ks4R2S4D#g1l_wH-nBOMONrDqOkeEvQ/folder/g5RRHJQC

hackpascal commented on target/linux/generic/files/drivers/mtd/nand/mtk_bmt.c:
```diff
@@ -431,6 +431,9 @@ int mtk_bmt_attach(struct mtd_info *mtd)
 	mtk_bmt_replace_ops(mtd);
 
+	of_property_read_u32(np, "mediatek,bmt-mtd-overridden-oobsize",
+				 &bmtd.mtd->oobsize);
```
It's better to add a check here. The bmtd.mtd->oobsize passed from fdt must be valid (<= real oobsize from lower mtd)

csharper2005 replied:
Done.

```
[    0.872478] spi-nand spi0.0: FORESEE SPI NAND was found.
[    0.877795] spi-nand spi0.0: 128 MiB, block size: 128 KiB, page size: 2048, OOB size: 64
[    0.886485] Signature found at block 1023 [0x07fe0000]
[    0.891663] NMBM management region starts at block 960 [0x07800000]
[    0.898719] First info table with writecount 0 found in block 960
[    0.907252] Second info table with writecount 0 found in block 963
[    0.913444] NMBM has been successfully attached
[    0.918349] 10 fixed-partitions partitions found on MTD device spi0.0
[    0.924831] Creating 10 MTD partitions on "spi0.0":
```

```
[    0.872005] spi-nand spi0.0: Winbond SPI NAND was found.
[    0.877323] spi-nand spi0.0: 128 MiB, block size: 128 KiB, page size: 2048, OOB size: 96
[    0.885547] NMBM: mtd OOB size has been overridden to 64B
[    0.891247] Signature found at block 1023 [0x07fe0000]
[    0.896376] NMBM management region starts at block 960 [0x07800000]
[    0.903543] First info table with writecount 1 found in block 960
[    0.911977] Second info table with writecount 1 found in block 963
[    0.918156] NMBM has been successfully attached
[    0.923076] 10 fixed-partitions partitions found on MTD device spi0.0
```

robimarko commented:
Thanks! Rebased on top of main and merged!

Ser9ei commented:
How it will work if vendor will release him router with new spi-nand where OOB will be different from 64 (and this value will be correct for the U-Boot and for the Linux kernel).
As for me, we can reduce the risk if define in the DTS id of a specific model (id of W25N01KV in our case with xiaomi ax3000t).

csharper2005 commented:
> How it will work if vendor will release him router with new spi-nand where OOB will be ‘wrong’ and different from 64.

The property is ignored when spi-nand driver `oobsize` is equal to `spare size` in nmbm signature.

csharper2005 commented:
> How it will work if vendor will release him router with new spi-nand where OOB will be ‘wrong’ and different from 64.

In this case will need to do autodetect instead of overriding.
